kafka-client/src/main/java/dev/responsive/kafka/internal/clients/AsyncStreamsConsumer.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2024 Responsive Computing, Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package dev.responsive.kafka.internal.clients;
+
+import static dev.responsive.kafka.internal.utils.Utils.extractThreadNameFromConsumerClientId;
+
+import dev.responsive.kafka.api.async.internals.AsyncThreadPoolRegistry;
+import java.time.Duration;
+import java.util.Map;
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.common.TopicPartition;
+
+/**
+ * Simple wrapper around the underlying ResponsiveConsumer that handles async stuff under ALOS.
+ * <p>
+ * The async consumer has two jobs: first, it registers and manages the lifecycle of the
+ * async thread pool associated with this StreamThread. Second, it's used to make sure any
+ * pending async records get flushed through the topology before committing offsets for
+ * those records.
+ * <p>
+ * If EOS is enabled, the async consumer will not be used at all and the above responsibilities
+ * fall instead to the {@link AsyncStreamsProducer}
+ */
+public class AsyncStreamsConsumer<K, V> extends DelegatingConsumer<K, V> {
+
+  private final String streamThreadName;
+  private final AsyncThreadPoolRegistry asyncThreadPoolRegistry;
+  private final Runnable flushAsyncProcessors;
+
+  public AsyncStreamsConsumer(
+      final Consumer<K, V> delegate,
+      final String clientId,
+      final AsyncThreadPoolRegistry asyncThreadPoolRegistry
+  ) {
+    super(delegate);
+    this.streamThreadName = extractThreadNameFromConsumerClientId(clientId);
+
+    this.asyncThreadPoolRegistry = asyncThreadPoolRegistry;
+    final var asyncThreadPoolRegistration = asyncThreadPoolRegistry
+        .startNewAsyncThreadPool(streamThreadName);
+    this.flushAsyncProcessors = asyncThreadPoolRegistration::flush;
+
+  }
+
+  @Override
+  public void commitSync(final Map<TopicPartition, OffsetAndMetadata> offsets) {
+    flushAsyncProcessors.run();
```

Comment:
I think the approach here breaks ALOS, because the producer has already been flushed at this point. I think we need to call producer.flush here.

Reply:
Good catch. Changed it to just do the async flushing inside/before the Producer#flush instead of the consumer/producer commitOffsets/commitTxn. Also simplified things so that the async threadpool lifecycle is always managed by the AsyncConsumer and the flushing is always managed by the AsyncProducer, and we always create both clients, regardless of eos.